packages/frontend/src/model/model_analyzer.tsx

```diff
@@ -0,0 +1,97 @@
+import { MultiProvider } from "@solid-primitives/context";
+import { type Accessor, Show, createContext, useContext } from "solid-js";
+import { Dynamic } from "solid-js/web";
+
+import type { DblModel } from "catlog-wasm";
+import {
+    type CellConstructor,
+    type FormalCellEditorProps,
+    NotebookEditor,
+    newFormalCell,
+} from "../notebook";
+import type { Theory } from "../theory";
+import { TheoryContext } from "./model_context";
+import type { ModelNotebookRef } from "./model_notebook_editor";
+import type { ModelAnalysis, ModelJudgment } from "./types";
+
+/** Notebook editor for analyses of models of double theories.
+ */
+export function ModelAnalyzer(props: {
+    handle: DocHandle<unknown>;
+    path: Prop[];
+    modelNotebookRef: ModelNotebookRef;
+}) {
+    return (
+        <MultiProvider
+            values={[
+                [TheoryContext, props.modelNotebookRef.theory],
+                [ModelContext, props.modelNotebookRef.model],
+                [ValidatedModelContext, props.modelNotebookRef.validatedModel],
+            ]}
+        >
+            <NotebookEditor
+                handle={props.handle}
+                path={props.path}
+                notebook={props.modelNotebookRef.modelNotebook().analysis}
+                changeNotebook={(f) =>
+                    props.modelNotebookRef.changeModelNotebook((model) => f(model.analysis))
+                }
+                formalCellEditor={ModelAnalysisCellEditor}
+                cellConstructors={modelAnalysisCellConstructors(props.modelNotebookRef.theory())}
+                noShortcuts={true}
+            />
+        </MultiProvider>
+    );
+}
+
+function ModelAnalysisCellEditor(props: FormalCellEditorProps<ModelAnalysis<unknown>>) {
+    const theory = useContext(TheoryContext);
+    const model = useContext(ModelContext);
+    const validatedModel = useContext(ValidatedModelContext);
+
+    return (
+        <Show when={theory?.()}>
+            {(theory) => (
+                <Show
+                    when={theory().getModelAnalysis(props.content.tag)}
+                    fallback={<span>Internal error: model view not defined</span>}
+                >
+                    {(meta) => (
+                        <Dynamic
+                            component={meta().component}
+                            model={model?.() ?? []}
+                            validatedModel={validatedModel?.() ?? null}
+                            theory={theory()}
+                            content={props.content.content}
+                            changeContent={(f: (c: unknown) => void) =>
+                                props.changeContent((content) => f(content.content))
+                            }
+                        />
+                    )}
+                </Show>
+            )}
+        </Show>
+    );
+}
+
+function modelAnalysisCellConstructors(theory?: Theory): CellConstructor<ModelAnalysis<unknown>>[] {
```

Comment:
This should just take in a theory: Theory; the case when theory is undefined should be handled elsewhere.